### PR TITLE
fern: learnable scalar/channel gate on surf→vol xattn residual

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The surf→vol cross-attention module (PR #823) injects surface geometry information into the volume decoder by adding the xattn output directly to the volume hidden state as a residual. However, this **fixed additive injection** has no mechanism to modulate *how much* the surface geometry signal is incorporated — it's always added at full strength from the first training step.

The hypothesis is that a **learnable scalar gate on the xattn residual** will improve results by:
1. Initialising at zero (so training starts identically to the non-xattn baseline — the model inherits the pre-xattn optimum and learns to use geometry conditioning incrementally)
2. Allowing the model to learn the optimal mixing ratio between the backbone's own volume representations and the cross-attention geometry signal

This is analogous to the zero-init gating used in ControlNet and ResNet-style residual gating. The gate avoids destabilising early training and gives the optimiser a smooth path to the optimal xattn contribution.

Two arms test the gating mechanism design:
- **Arm A**: Scalar gate `α` (single learnable scalar, initialized to 0): `output = volume_hidden + α * xattn_out`
- **Arm B**: Per-head gate `α` (learnable vector of size `num_heads=4`, initialized to zeros): `output = volume_hidden + (α.unsqueeze(-1) * xattn_out_per_head).sum(dim=head)`

Both arms use the full SOTA xattn config (PR #823) as the base — same architecture, same hyperparameters, only the residual injection changes.

## Instructions

**Step 1: Implement the learnable gate in the surf→vol xattn module**

Find the `SurfToVolCrossAttention` module (or equivalent) in `target/model.py` (or wherever `--use-surf-to-vol-xattn` is implemented). It currently does something like:
```python
xattn_out, _ = self.cross_attn(query=vol_hidden, key=surf_hidden, value=surf_hidden)
vol_hidden = vol_hidden + xattn_out  # fixed residual
```

**Arm A — scalar gate:**
Add `self.xattn_gate = nn.Parameter(torch.zeros(1))` to the module's `__init__`. Change the residual to:
```python
xattn_out, _ = self.cross_attn(query=vol_hidden, key=surf_hidden, value=surf_hidden)
vol_hidden = vol_hidden + self.xattn_gate * xattn_out
```

**Arm B — per-head gate:**
Add `self.xattn_gate = nn.Parameter(torch.zeros(num_heads))` to `__init__`. Reshape the xattn output before gating so each head's contribution is scaled independently, then sum. (The `nn.MultiheadAttention` in PyTorch already combines heads in the forward pass — you'll need to use `batch_first=True` and split the output by head, or alternatively override `out_proj` to apply the gate before the output projection.) A simpler approach that approximates per-head gating: apply the gate as a learned linear re-weighting of the output channels: `self.xattn_gate = nn.Parameter(torch.zeros(embed_dim))` and `vol_hidden = vol_hidden + self.xattn_gate * xattn_out` (element-wise per-channel scalar).

Add a CLI flag `--xattn-gate-type` with choices `['none', 'scalar', 'channel']` (default `'none'` preserves the original behaviour). Set `'scalar'` for Arm A and `'channel'` for Arm B.

Log `self.xattn_gate.mean().item()` to W&B as `xattn_gate_mean` each epoch to track how the gate evolves during training.

**Step 2: Run two arms**

Use `--wandb-group fern-xattn-gate` to group both arms.

**Arm A — scalar gate:**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --xattn-gate-type scalar \
  --wandb-group fern-xattn-gate \
  --wandb-name fern/xattn-gate-scalar
```

**Arm B — channel gate:**
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent fern \
  --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --xattn-gate-type channel \
  --wandb-group fern-xattn-gate \
  --wandb-name fern/xattn-gate-channel
```

**Step 3: Kill gates (per arm)**
- EP1: val_abupt < 30%
- EP2: val_abupt < 16%
- EP3: val_abupt < 8%
- EP4: val_abupt ≤ 6.4407% (must beat single-model SOTA)

**Step 4: Report**

In your PR results comment, include:
- For each arm: final val_abupt, test_abupt, test_vol_p, best epoch
- The final learned value of `xattn_gate_mean` at the best epoch for each arm (this tells us whether the gate converged to a meaningful non-zero value)
- The W&B run IDs for both arms

## Baseline

Current single-model SOTA (PR #823, run `ghh0s4ne`):
- val_abupt: **6.4407%**
- test_abupt: **7.6992%**
- test_vol_p: **11.6704%**

Kill gate: EP4 val_abupt ≤ 6.4407%

| Metric | SOTA |
|--------|------|
| val_abupt | 6.4407% |
| test_abupt | 7.6992% |
| test_vol_p | 11.6704% |

Reproduce SOTA (PR #823):
```bash
cd target/ && torchrun --standalone --nproc_per_node=8 train.py \
  --agent nezuko --optimizer lion --lr 9e-5 --weight-decay 5e-4 \
  --tau-y-loss-weight 1.5 --tau-z-loss-weight 2.0 --surface-loss-weight 2.0 \
  --ema-decay 0.999 --grad-clip-norm 0.5 --lr-warmup-epochs 1 \
  --pos-encoding-mode string_separable --use-qk-norm \
  --rff-num-features 16 --rff-init-sigmas "0.25,0.5,1.0,2.0,4.0" \
  --lr-cosine-t-max 13 --epochs 13 \
  --vol-points-schedule "0:16384:3:32768:6:49152:9:65536" \
  --no-compile-model \
  --model-layers 5 --model-hidden-dim 512 --model-heads 4 --model-slices 128 \
  --batch-size 4 --validation-every 1 \
  --train-surface-points 65536 --eval-surface-points 65536 \
  --train-volume-points 65536 --eval-volume-points 65536 \
  --use-surf-to-vol-xattn \
  --wandb-group nezuko-surf-vol-xattn \
  --wandb-name nezuko/surf-vol-xattn-13ep
```

### CODE REVIEW

**Comment by morganmcg1** [comment]
STUDENT fern: Implementation pushed (commit 35c44df). Arm A launched.

**Implementation summary**

Added `--xattn-gate-type {none,scalar,channel}` to `SurfaceTransolver` (model.py) and `Config` (train.py):
- `none` (default): preserves PR #823 behaviour — zero-init out_proj + post-add LayerNorm
- `scalar` (Arm A, ReZero-style): `α = Param(zeros(1))`, residual = h + α·xattn_out, **standard PyTorch init for out_proj** (NOT zeros)
- `channel` (Arm B, LayerScale-style): `α = Param(zeros(n_hidden))`, residual = h + α·xattn_out, **standard init for out_proj**

**Critical fix vs naive impl**: when both α=0 AND out_proj=zeros, the gradient `dL/dα = dL/dout · xattn_out = 0` and `dL/dW_o = α · (...) = 0` — both are stuck at zero forever. ReZero/LayerScale rely on the sublayer being standard-initialized so `F(x) ≠ 0` and α gets a real gradient on step 1. I therefore drop the zero-init of `out_proj.weight` only when the gate is enabled. Verified by autograd: scalar gate.grad = -1.3e-3 at step 1; channel gate.grad nonzero on all 64/64 dims of the test model.

Logs `xattn_gate_mean` (and abs_mean/std/min/max) to W&B per epoch via `collect_xattn_gate_metrics`.

**Arm A — scalar gate**

- W&B: https://wandb.ai/wandb-applied-ai-team/senpai-v1-drivaerml-ddp8/runs/gutn4sme
- Pace: 1.91 it/s (matches PR #823 SOTA), Epoch 1 ETA ≈ 1h33m, end-to-end ETA ≈ 14.7h
- Launched with `SENPAI_TIMEOUT_MINUTES=1200` (20h budget) following PR #823 precedent
- Arm B (channel) will launch after Arm A finishes — total ≈ 30h sequential (8 GPUs each, per PR spec)

**Comment by morganmcg1** [comment]
ADVISOR: EP1 KILL — Arm A (scalar gate).

EP1 val_abupt=31.34% fails the <30% gate by 1.34pp. Kill Arm A (run gutn4sme) now.

Despite the correct fix (drop zero-init of out_proj when gate enabled), the scalar ReZero gate did not provide early-training acceleration. The gate starts at α=0 meaning the xattn contribution is zeroed out at init, which means the vol tokens spend EP1 learning without any surface signal — the gate only becomes active once α drifts away from zero, which is slow for a scalar. Standard zero-init out_proj (PR #823 style) lets the full out_proj learn from step 1 even though the output starts at zero.

**Arm B (channel LayerScale):** Do NOT launch Arm B. Channel LayerScale has the same fundamental problem — α=zeros(n_hidden) means zero xattn contribution at init, just at channel granularity. The bottleneck is EP1 performance, and the gate mechanism itself is the issue, not the gate dimensionality.

**Instead, for your next experiment:** I will assign you a fresh hypothesis. Close this PR once Arm A is confirmed killed.

**Comment by morganmcg1** [comment]
STUDENT fern: Arm A (run `gutn4sme`) killed per advisor instructions.

- Training process (PID 1246011) was terminated at 2026-05-09T05:25:33Z — confirmed all 8 GPU workers exited and GPU memory is back to 0 MiB.
- Run had progressed to ~step 2067/10864 of EP2 (~19% through EP2) when killed; EP1 val_abupt=31.34% already failed the <30% gate.
- Arm B (channel) was **not** launched per advisor instructions.
- Closing this PR now; awaiting fresh hypothesis assignment.
